docs/rfc/rfc-12.adoc

```diff
@@ -0,0 +1,430 @@
+challenge. Having a small set of maintainers (or even a single one)
+amplifies the aforementioned SPoF problem.
+
+Last but not least, such a design may raise some censorship-related concerns.
+Although authorized maintainers are purely a coordination mechanism and
+cannot force wallets to do anything against the protocol rules, they can just
+become idle on purpose and force refunds on specific deposits or timeouts
+on redemptions.
+
+This is why we need to replace the `WalletCoordinator`-based mechanism with
+a more robust and decentralized alternative.
+
+== Proposal
+
+Here we propose to replace the global `WalletCoordinator`-based wallet
+coordination with an in-wallet mechanism, where each wallet is responsible for
+its own coordination.
+
+=== Goal
+
+The goal of this proposal is to eliminate the current naive model, based
+on the `WalletCoordinator` contract and authorized maintainers. Pushing the
+responsibility of coordination onto wallets should solve the problems of the
+current mechanism and make wallet coordination reliable, decentralized,
+and censorship-resistant.
+
+=== Implementation
+
+The implementation must meet the proposal goals and be neutral for system
+stakeholders as much as possible. Specifically, the implementation:
+
+- Should be possible to introduce in a live system
+- Should not have negative impact on key architecture attributes
+  (especially security, reliability, maintainability, and performance)
+- Should involve only wallet signers and leverage secure broadcast channels
+  for inter-signer communication, just as for other wallet actions
+
+As mentioned in the introduction of this RFC, the original consensus-based
+solution does not meet the second requirement, as it negatively impacts
+reliability and performance. The proposed implementation is something
+between the current `WalletCoordinator`-based model and a consensus-based
+mechanism. The following sections describe it in detail.
+
+==== Drop the `WalletCoordinator` contract
+
+First and foremost, wallet signers must stop relying on the `WalletCoordinator`
+contract. The wallet client software should no longer listen to events emitted
+from it and should not trigger any actions as a result.
+
+At this point, it is worth mentioning that the `WalletCoordinator` contract
+provides a mechanism of wallet locks, which ensures that an individual wallet
+receives only one proposal at a time, and a fixed time delay is preserved
+before a subsequent one can be submitted. Getting rid of the `WalletCoordinator`
+contract means that the wallet client implementation must handle that problem
+on its own. Fortunately, this is already covered in the wallet client
+implementation living in the `keep-core` repository. Specifically:
+
+- The `walletDispatcher` component ensures that only one action at a time
+is executed by the given wallet
+- The `EnsureWalletSyncedBetweenChains` function ensures that previous wallet
+actions were properly completed
```

Comment:
The mechanism in `EnsureWalletSyncedBetweenChains` does not guarantee the result is 100% accurate and I don't think there is a way to guarantee it. For example, there is no way to be sure a new transaction will not land into the mempool a few milliseconds after [this check](https://github.com/keep-network/keep-core/blob/324f66fb3f1003f6cfeb7d4149ae3f1d902dba2e/pkg/tbtc/wallet.go#L531). The fact the wallet acts only once, at the end of the coordination window helps to address some problems but I do wonder if we shouldn't add some idleness buffer between the coordination window end and the time validations are triggered.

For example, a wallet retries sweep transaction signing multiple times and succeeds just before the coordination window ends. A malicious leader observe that although the signing succeeded, the transaction is not yet in the mempool. A malicious leader proposes another signing, and it passes given an unfortunate timing, leading to another, conflicting signing.

Reply:
Side note, since we are talking about `EnsureWalletSyncedBetweenChains`: shouldn't we check the mempool as well in the scenario [when the main UTXO exists](https://github.com/keep-network/keep-core/blob/324f66fb3f1003f6cfeb7d4149ae3f1d902dba2e/pkg/tbtc/wallet.go#L489)?

Reply:
Just to make sure this is clear (let me know if it's not): a coordination window is only for determining what action should be executed. The end of the coordination window is the start of the wallet action (validation + signing + broadcasting). 

Of course, a wallet action can overlap with the next coordination window. This is a normal case and I believe this is the concern you are raising with the mentioned example. I think we are protected because:

The majority executing the protracted action have this fact reflected in the `walletDispatcher`. This component will not allow to start a new action before the previous one is completed. Our current action implementations exit with success only when the resulting BTC transaction [lands in the mempool](https://github.com/keep-network/keep-core/blob/324f66fb3f1003f6cfeb7d4149ae3f1d902dba2e/pkg/tbtc/wallet.go#L326). Even if a leader manage to start a new action right after the previous one completes (this is very hard to achieve in practice), `EnsureWalletSyncedBetweenChains` will be able to catch the state drift as the BTC transaction being result of the previous action already lives in the mempool.

Regardless, adding a short idleness buffer may be a good idea. It may allow slower members to catch up and prepare for wallet action execution. Of course, it slightly pushes back the action execution and increases the chance of overlap with the next coordination window. I also think the coordination layer implementation must be aware of the wallet state to not naively schedule wallet actions that will be certainly nuked at the action validation stage.

Reply:
> Side note, since we are talking about EnsureWalletSyncedBetweenChains: shouldn't we check the mempool as well in the scenario [when the main UTXO exists](https://github.com/keep-network/keep-core/blob/324f66fb3f1003f6cfeb7d4149ae3f1d902dba2e/pkg/tbtc/wallet.go#L489)?

This is already covered. If the main UTXO is spent by a transaction that is still in the mempool this will be detected as well. See [this comment](https://github.com/keep-network/keep-core/blob/324f66fb3f1003f6cfeb7d4149ae3f1d902dba2e/pkg/tbtc/wallet.go#L481).

Reply:
> Just to make sure this is clear (let me know if it's not): a coordination window is only for determining what action should be executed. The end of the coordination window is the start of the wallet action (validation + signing + broadcasting).

> Of course, a wallet action can overlap with the next coordination window. This is a normal case and I believe this is the concern you are raising with the mentioned example. 

Correct, this is all clear and I think this mechanism makes a lot of sense. We are on the same page here. 💯 

> The majority executing the protracted action have this fact reflected in the walletDispatcher. This component will not allow to start a new action before the previous one is completed. Our current action implementations exit with success only when the resulting BTC transaction [lands in the mempool](https://github.com/keep-network/keep-core/blob/324f66fb3f1003f6cfeb7d4149ae3f1d902dba2e/pkg/tbtc/wallet.go#L326). Even if a leader manage to start a new action right after the previous one completes (this is very hard to achieve in practice), `EnsureWalletSyncedBetweenChains` will be able to catch the state drift as the BTC transaction being result of the previous action already lives in the mempool.

My understanding of the mechanism referenced [here](https://github.com/keep-network/keep-core/blob/324f66fb3f1003f6cfeb7d4149ae3f1d902dba2e/pkg/tbtc/wallet.go#L326) is that we wait for the transaction to be in a mempool, at minimum. We do not require even a single confirmation. This - connected with the fact the wallet is locked in the action dispatcher - would be enough if we could trust the mempool that the transaction will not disappear from there. 

I am not familiar enough with the mechanisms of mempool and how Electrum implementations scan the mempool to guarantee we can rely on the mempool check queries. In practice, this check is probably enough for 99.999% of cases but I do think we could better protect against the remaining 0.0001% by on of those options:

1. Instead of waiting for the transaction to be in a mempool, block the wallet until it has at least one confirmation. In the future, we'll have to plug the fee increase mechanism into it. That fee mechanism could work without a coordinator. For example, we automatically increase the fee x2 after 30 minutes of waiting.

    Disclaimer: I haven't validated this idea with the code so I am not 100% sure if there are some negative implications.

3. Do not accept messages from the leader if less than N blocks are left before the window ends. This is similar to the previous idea but I think it is potentially better given in the vast majority of cases the leader should propose something at the beginning of the coordination window and this approach has a lower possibility of actions overlapping in the next coordination window compared to the previous proposition.

Reply:
> I am not familiar enough with the mechanisms of mempool and how Electrum implementations scan the mempool to guarantee we can rely on the mempool check queries

The given transaction can quickly disappear from the mempool only if it is invalid (e.g. double spending takes place). Otherwise, it lives in the mempool until it gets mined or evicted if the fee is significantly too low. Time to eviction depends on node implementation but the rule of thumb is that the transaction lasts at least for ~48 hours.

> In practice, this check is probably enough for 99.999% of cases but I do think we could better protect against the remaining 0.0001% by one of those options:

Yep, the risk that a transaction gets evicted from the mempool immediately after being submitted is negligible hence this check is safe right now. However, I do agree it makes sense to introduce an additional protection just in case.

I think we should use the second proposition. I think the first one breaks our implicit rule of thumb that: "the wallet action completes once the transaction is submitted; coordination happens beyond". This rule allows to keep action implementations simple and not extend their execution time beyond the absolutely necessary range. RBF is a good example by the way. From the wallet's perspective, this is a completely separate action as it requires executing signing on different data, this is not something the wallet can do in a quick Ethereum-like fashion. Squashing it with action execution would make the action implementation way more complicated. It's better to outsource that complexity to the coordination layer.

Returning back to the second proposition I think it makes sense to make followers await leader's message just for X% of the coordination window. The remaining 100-X% should be a buffer allowing to validate the message and prepare for action execution. It will be also an implicit additional protection against the case mentioned in this thread. I think X should be 80%. With 100 blocks of window duration, that gives 80 blocks (~16 minutes) for communication and 20 blocks (~4 minutes) for the buffer. What do you think?

Reply:
> The given transaction can quickly disappear from the mempool only if it is invalid (e.g. double spending takes place). Otherwise, it lives in the mempool until it gets mined or evicted if the fee is significantly too low. Time to eviction depends on node implementation but the rule of thumb is that the transaction lasts at least for ~48 hours.

Right, this is what we observed and maybe even confirmed in the code some time ago. But can we guarantee this will always be the case? I am afraid we can't. 

> Returning back to the second proposition I think it makes sense to make followers await leader's message just for X% of the coordination window. The remaining 100-X% should be a buffer allowing to validate the message and prepare for action execution. It will be also an implicit additional protection against the case mentioned in this thread. I think X should be 80%. With 100 blocks of window duration, that gives 80 blocks (~16 minutes) for communication and 20 blocks (~4 minutes) for the buffer. What do you think?

Agreed! 💯

Reply:
> But can we guarantee this will always be the case? I am afraid we can't.

Sure, this is why we agreed to add the additional protection.

> Agreed! 💯

See: 6432ff0f